### ISSUE TITLE
Recursive loop when nesting callSuper() (call stack exceeded)

### ISSUE THREAD

**Issue by MagoDopado** [CocktailJS/cocktail#26]
Got myself in some trouble when trying to extend a few classes. 
And since the proof is in eating the pudding, here's some code:

* Abstract class Algorithm
```javascript
cocktail.mix({
	'@exports': module,
	'@as': 'class',
	'@traits': [AlgorithmInterface],
	'@logger' : [console, "Algorithm Base:"],

	constructor: function() {
	  //Should be initialized by some especification of this class.
	  this._victims = undefined;
	  this._requirements = undefined;
	  this._finalized = new Queue();
	  this._filters = [];
	},

	initialize: function(requirements) {
	  this._requirements = requirements;
	  this._finalized = new Queue();
	}
});
```

* Concrete base class Fifo
```javascript
cocktail.mix({
	'@exports': module,
	'@as': 'class',
	'@extends': Algorithm,
	'@traits': [AlgorithmInterface],
	'@logger' : [console, "Algorithm Fifo:"],

	constructor: function() {
	  this.callSuper("constructor");
	  this._victims = new Queue();
	  this.log("Created.");
	},

	initialize: function(requirements) {
	  this.callSuper("initialize", requirements);
	  this._victims = new Queue();
	}
});
```

* Most derived class, LRU
```javascript
cocktail.mix({
	'@exports': module,
	'@as': 'class',
	'@extends': Fifo,
	'@traits': [AlgorithmInterface],
	'@logger' : [console, "Algorithm Lru:"],

	constructor: function() {
	  this.callSuper("constructor");
	  this._victims = new ReQueueQueue();
	  this.log("Created.");
	},

	initialize: function(requirements) {
	  this.callSuper("initialize", requirements);
	  this._victims = new ReQueueQueue();
	}
});
```

As you can see, there are 2 chains of callSuper methods, one in the constructor call, the other in the initialize method call, but the call only starts looping when initialize is called.

**Comment by brunocascio** [comment]
In addition of @MagoDopado, the output error is `RangeError: Maxium call stak exceeded`

**Comment by elmasse** [comment]
Thanks for reporting this. I'll take a look as soon as I get back later this week. Probably around Thursday.

**Comment by elmasse** [comment]
@brunocascio @MagoDopado this issue should be fixed in the latest commit. Please try it by using the master branch in npm dependency and let me know if you have any issue. I'll publish a new version as soon as you confirm this is working properly.

```json
{
   "dependencies": {
    "cocktail": "git://github.com/CocktailJS/cocktail.git"
  }
}

````

**Comment by MagoDopado** [comment]
Works fine for the case described above, but won't work in a different setup.
I'll explain further:

* Abstract class Algorithm
```javascript
cocktail.mix({
    '@exports': module,
    '@as': 'class',
    '@traits': [AlgorithmInterface],
    '@logger' : [console, "Algorithm Base:"],

    constructor: function() {
      //Should be initialized by some especification of this class.
      this._victims = undefined;
      this._requirements = undefined;
      this._finalized = new Queue();
      this._filters = [];
    },

    initialize: function(requirements) {
      this._requirements = requirements;
      this._finalized = new Queue();
    },

    update: function(requirement) {
      this._finalized.add(requirement);
    }
});
```
* Concrete base class Fifo
```javascript
cocktail.mix({
    '@exports': module,
    '@as': 'class',
    '@extends': Algorithm,
    '@traits': [AlgorithmInterface],
    '@logger' : [console, "Algorithm Fifo:"],

    constructor: function() {
      this.callSuper("constructor");
      this._victims = new Queue();
      this.log("Created.");
    },

    initialize: function(requirements) {
      this.callSuper("initialize", requirements);
      this._victims = new Queue();
    },

    update: function(requirement) {
      this.callSuper("update", requirement);
      this._victims.add(requirement);
    }
});
```
* Most derived class, LRU
```javascript
cocktail.mix({
    '@exports': module,
    '@as': 'class',
    '@extends': Fifo,
    '@traits': [AlgorithmInterface],
    '@logger' : [console, "Algorithm Lru:"],

    constructor: function() {
      this.callSuper("constructor");
      this._victims = new ReQueueQueue();
      this.log("Created.");
    },

    initialize: function(requirements) {
      this.callSuper("initialize", requirements);
      this._victims = new ReQueueQueue();
    }
});
```

As we can see in this extended example (sorry I didn't add it before, I thought it was the same case but it's not) we have 3 diferent types of chains.

* ``constructor()``
  * **Explanation**: This is a simple chain of constructors. All classes define a constructor and the derived ones call to super.
  * **Before**: I make a special case for this because it never looped, went a litle deep in the call, and when creating an LRU object, FIFO constructor never was called but Algorithm constructor did.
  * **How is it working now?** It properly calls all constructors in its chain.
* ``initialize()``
 * **Explanation**: Chain of initialize methods . All classes define an initialize method and the derived ones call to super.
  * **Before**:  This was the chain that looped. LRU's initialize would call FIFO's initialize that in the context of LRU would resolve in calling the FIFO's update again and again.
  * **How is it working now?** It properly calls all initialize methods in its chain.
* ``update()``
  * **Explanation**: Chain of update methods . Only Algorithm & FIFO implement the update method and LRU inherits it from FIFO. FIFO's update method calls to super.
 * **Before**:  This chain looped too.
 * **How is it working now?** Calls FIFO's update method. When super is called, super is resolved as the FIFO class so FIFO's update method is called again. Then super is resolved as Algorithm & Algorithm's update method is called. This beahavior is not the expected since FIFO's update method is called twice.
 * *__This behavior can be easily be fixed defining the update method in LRU class like this__*
 ```javascript
 update: function (requirement) {
    return this.callSuper("update", requirement);
 }
 ```
 
 Thanks for the quick answer!

**Comment by elmasse** [comment]
@MagoDopado @brunocascio I have added a bunch of test scenarios including this one. Latest commit should fix them. Please let me know how it goes!

Thanks!
Max

**Comment by MagoDopado** [comment]
All call's work perfectly! Thanks a lot. :+1:

**Comment by elmasse** [comment]
Awesome! 
New version has been released (0.7.2) on npm which contains all these fixes.

**Comment by brunocascio** [comment]
Thanks! @elmasse
